language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -969,12 +1085,125 @@ fn pkg_deps_equal<'a>(
     Vec<(&'a Dependency, PM::PackageName)>,
     Vec<(&'a Dependency, PM::PackageName)>,
 ) {
-    let pkg_edges = BTreeSet::from_iter(pkg_graph.edges(pkg_name).map(|(_, pkg, dep)| (dep, pkg)));
+    // consider only finalized edges so that we don't include edges that otherwise should be missing
+    // from the graph
+    let pkg_edges = BTreeSet::from_iter(pkg_graph.edges(pkg_name).filter_map(|(_, pkg, dep)| {
+        if dep.finalized {
+            Some((dep, pkg))
+        } else {
+            None
+        }
+    }));
+
     let other_edges =
-        BTreeSet::from_iter(other_graph.edges(pkg_name).map(|(_, pkg, dep)| (dep, pkg)));
+        BTreeSet::from_iter(other_graph.edges(pkg_name).filter_map(|(_, pkg, dep)| {
+            if dep.finalized {
+                Some((dep, pkg))
+            } else {
+                None
+            }
+        }));
 
     let (pkg_deps, other_deps): (Vec<_>, Vec<_>) = pkg_edges
         .symmetric_difference(&other_edges)
         .partition(|dep| pkg_edges.contains(dep));
     (pkg_deps, other_deps)
 }
+
+/// In presence of dependency overrides, returns a "winning" package if any (in which case it
+/// returns an error). The algorithm relies on the fact that this check is performed when a new
+/// package is added to the graph already containing package with the same name - this allows us to
+/// resolve the conflicts pair-wise even if ultimately a package with a given name is a dependency
+/// of more than two other packages.
+///
+/// First we check if package already existing in the graph has been inserted there as a result of
+/// an override. If it was, then either:
+/// - the new package is dominated by the same override
+/// or
+/// - no valid override can be found and error must be reported
+///
+/// Because the overridden dependencies are processed first for each package, it is guaranteed that
+/// when searching for overrides for a child (new package), a correct override will be found if it
+/// exists because by construction it must be already present in the graph.
+///
+/// If package already existing in the graph has not been inserted as a result of an override then
+/// either:
+/// - the new package has been inserted as a result of an override and:
+///     - it's dominating override for the existing package then new package is the winner
+///     or
+///     - it's not a dominating override for the existing package then an error must be reported
+/// or
+/// - the new package has not been inserted as a result of an override and an error must be reported
+///
+/// Correctly finding dominators only works because when trying to add a conflicting package through
+/// a new dependency edge, we add this edge early in an non-finalized state (before the package is
+/// fully processed), finalizing the edges only after package processing is done.
```

Comment:
Additionally, now I've read the full scope of changes, I'm a little concerned that adding finalized edges may break a legitimate dominating override.  The example I have in mind goes like this:

```
     ┌───R───┐
     │   │   │
     │   │o  │
     ▼   ▼   ▼
 ┌───C   B◄──A
 │   │       │
o│   │       │
 ▼   ▼       │
 D◄──E◄──B'◄─┘
     │
     │
     ▼
     D'
```

This is a legitimate dependency graph where `R` override's `A`'s dependency on `B` (moving it from `B'` to `B`), and `C` overrides `E`'s dependency on `D` similarly.  But if you include non-finalized edges when computing dominators, you get into a situation where `C`'s override of `D` is not recognised because `C` does not dominate `E` (there's a separate path to `E` via the non-finalized `A -> B'` edge).